Mne v0.15

ckiefer0 opened the pull request:
interpolate_bads v3 (June 5th, 2018) by Eberhard
ica_on_unfiltered by me

pravsripad commented:
@ckiefer0 as discussed, we can remove the mne ICA class and simply use the function (ica_on_unfiltered) with the modification that will allow us to apply ICA on unfiltered data keeping the mean and standard deviation intact. 

could you also please add two simple examples using some sample data so that people can start using your functions quickly?

pravsripad commented on jumeg_interpolate_bads.py:
```diff
@@ -0,0 +1,117 @@
+    info_from = pick_info(inst.info, picks_good)
+    info_to = pick_info(inst.info, picks_bad)
+
+    if origin is None:
```
@ebeich are these changes 4D specific or would it work also for data from other systems?

ckiefer0 commented:
I added two examples, one for interpolate_bads and one for apply_ica_replace_mean_std.

I forgot that there is a problem with how mne.preprocessing.ICA whitens the data. I fixed that in my version of mne but others will not have that change. If you run the example you should see that all the entries for the standard deviation are actually the same in the case of the "original" ica.apply. This is why I recommend using apply_ica_replace_mean_std instead of ica.apply. (This is also the reason why I renamed this from ica_on_unfiltered.)

pravsripad commented:
Thanks @ckiefer0 for the changes. I tried running the example and get the below error,

```
psripad@pravmac:~/mne/jumeg_tests $python apply_ica_replace_mean_std.py
Successfully extracted to: [u'/Users/psripad/mne_data/MNE-sample-data']
Opening raw data file /Users/psripad/mne_data/MNE-sample-data/MEG/sample/sample_audvis_raw.fif...
    Read a total of 3 projection items:
        PCA-v1 (1 x 102)  idle
        PCA-v2 (1 x 102)  idle
        PCA-v3 (1 x 102)  idle
    Range : 25800 ... 192599 =     42.956 ...   320.670 secs
Ready.
Current compensation grade : 0
Reading 0 ... 166799  =      0.000 ...   277.714 secs...
Setting up band-pass filter from 1 - 45 Hz
l_trans_bandwidth chosen to be 1.0 Hz
h_trans_bandwidth chosen to be 11.2 Hz
Filter length of 1982 samples (3.300 sec) selected
[Parallel(n_jobs=2)]: Done 228 tasks      | elapsed:    1.9s
[Parallel(n_jobs=2)]: Done 366 out of 366 | elapsed:    2.9s finished
Fitting ICA to data using 305 channels.
Please be patient, this may take some time
Inferring max_pca_components from picks.
Selection by number: 60 components
Traceback (most recent call last):
  File "apply_ica_replace_mean_std.py", line 41, in <module>
    n_pca_components=None)
  File "/Users/psripad/miniconda2/lib/python2.7/site-packages/jumeg/decompose/ica_replace_mean_std.py", line 111, in apply_ica_replace_mean_std
    raise ValueError('Data input must be of Raw or Epochs type')
ValueError: Data input must be of Raw or Epochs type
```

what version of mne are you using?

pravsripad commented:
@ckiefer0 it is working now, it was a problem with my environment.

ckiefer0 commented:
Ok, good! I am using mne v0.15 btw

pravsripad commented on decompose/ica_replace_mean_std.py:
```diff
@@ -0,0 +1,213 @@
+
+    This method ensures that the correct mean and standard deviation are used for
+    whitening the input data.
+
```
Maybe we can also include the use case of the function here i.e.

USE CASE: Sometimes it is best to fit ICA on filtered and cleaned data in order to improve the quality of decomposition. The fitted ICA object cannot be directly applied on unfiltered data due to changes in the mean and standard deviation between the filtered and unfiltered data. This function can be used in this case. It computes the correct mean and std values from given unfiltered data and includes it in the ica object.

pravsripad commented on examples/apply_ica_replace_mean_std.py:
```diff
@@ -0,0 +1,59 @@
+pca_mean_replaced_unfilt = ica.pca_mean_
+pca_pre_whitener_replaced_unfilt = ica._pre_whitener
+
+# compare filtered and unfiltered data
```
We can only show the first 10 values or so of mean and std between filt and unfilt for clarity. Avoids the for loops.

pravsripad commented on examples/do_apply_ica_replace_mean_std.py:
```diff
@@ -53,7 +53,11 @@
 for idx in range(0, len(pca_pre_whitener_filt)):
     print pca_pre_whitener_filt[idx], pca_pre_whitener_replaced_filt[idx], pca_pre_whitener_replaced_unfilt[idx]
+    if idx >= 9:
+        break
```
simply doing below would work.. 
```
print pca_pre_whitener_filt[:10]
```
also, nice padding :)

ckiefer0 replied:
print pca_pre_whitener_filt[:10], print pca_pre_whitener_replaced_filt[:10], print pca_pre_whitener_replaced_unfilt[:10]
If I use this, wouldn't all the entries be plotted below each other? I did it this way because I wanted the three corresponding entries in one line.

pravsripad replied:
Ah okay, then this is fine 👍

pravsripad commented:
@ckiefer0 let me know when it's ready to be merged.

ckiefer0 commented:
I think I implemented all the changes now.

pravsripad commented:
Merging into master. Many thanks @ckiefer0 @ebeich .
